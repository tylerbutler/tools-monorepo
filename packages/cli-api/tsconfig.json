{
	"extends": "../../config/tsconfig.strict.json",
	"include": ["src/**/*"],
	"compilerOptions": {
		"rootDir": "./src",
		"outDir": "./esm",
	},
	"ts-node": {
		"esm": true,
	},
}
