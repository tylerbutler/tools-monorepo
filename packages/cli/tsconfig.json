{
	"extends": "../../config/tsconfig.strict.json",
	"include": ["src/**/*"],
	"compilerOptions": {
		"rootDir": "src",
		"outDir": "esm",
	},
	"ts-node": {
		"esm": true,
	},
	"references": [
		{
			"path": "../cli-api",
		},
		{
			"path": "../dill",
		},
	],
}
