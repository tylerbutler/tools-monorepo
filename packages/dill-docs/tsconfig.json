{
	"extends": "astro/tsconfigs/strictest",
	"references": [
		{
			"path": "../dill",
		},
	],
}
