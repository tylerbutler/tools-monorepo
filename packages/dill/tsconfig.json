{
	"extends": "../../config/tsconfig.strict.json",
	"include": ["src/**/*"],

	"compilerOptions": {
		"composite": true,
		"declaration": true,
		"declarationMap": true,
		"esModuleInterop": true,
		"incremental": true,
		"inlineSources": true,
		"jsx": "react",
		"lib": ["ES2020", "DOM", "DOM.Iterable"],
		// "noImplicitAny": false,
		// "noUnusedLocals": true,
		"pretty": true,
		"sourceMap": true,
		"strict": true,
		// "target": "ES2020",

		"types": ["node"],
		"noImplicitAny": true,
		"noUnusedLocals": false,
		"target": "ES2022",
		"noUncheckedIndexedAccess": false,
		"exactOptionalPropertyTypes": false,

		"rootDir": "src",
		"outDir": "esm",
		// TODO: Should this setting be disabled in the strict config?
		"noPropertyAccessFromIndexSignature": false,
	},
	"ts-node": {
		"esm": true,
	},
}
