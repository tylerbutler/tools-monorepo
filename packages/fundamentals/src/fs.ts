import { access } from "node:fs/promises";

/**
 * Check if a file or directory exists.
 *
 * @remarks
 * This is an async replacement for the deprecated `fs.exists()` and synchronous `fs.existsSync()`.
 * It uses `fs.promises.access()` internally to check file accessibility.
 *
 * @param path - The path to check for existence
 * @returns A promise that resolves to `true` if the path exists, `false` otherwise
 *
 * @example
 * ```typescript
 * if (await exists('./myfile.txt')) {
 *   console.log('File exists');
 * }
 * ```
 *
 * @public
 */
export async function exists(path: string): Promise<boolean> {
	try {
		await access(path);
		return true;
	} catch {
		return false;
	}
}
