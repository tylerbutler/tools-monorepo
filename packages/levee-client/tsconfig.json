{
	"$schema": "https://json.schemastore.org/tsconfig",
	"extends": "../../config/tsconfig.strict.json",
	"include": ["src/**/*"],
	"exclude": ["src/test/**/*"],
	"compilerOptions": {
		"rootDir": "./src",
		"outDir": "./esm",

		"exactOptionalPropertyTypes": false,
	},
}
