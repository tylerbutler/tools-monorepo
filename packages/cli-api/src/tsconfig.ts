import { readFileSync, writeFileSync } from "node:fs";
import detectIndent from "detect-indent";
import { sortJsonc } from "sort-jsonc";
import JSONC from "tiny-jsonc";
import type { TsConfigJson } from "type-fest";

/**
 * Sorting order for keys in the compilerOptions section of tsconfig. The groups and the order within each group are
 * based on the order at https://www.typescriptlang.org/tsconfig#compiler-options. However, the order of the groups has
 * been adjusted, and a few properties are moved earlier in the order since they're more important to our repo
 * tsconfigs.
 */
const compilerOptionsOrder = [
	"rootDir", // From the Modules group
	"outDir", // From the Emit group
	"module", // From the Modules group
	"moduleResolution", // From the Modules group

	// Emit
	"declaration",
	"declarationDir",
	"declarationMap",
	"downlevelIteration",
	"emitBOM",
	"emitDeclarationOnly",
	"importHelpers",
	"importsNotUsedAsValues",
	"inlineSourceMap",
	"inlineSources",
	"mapRoot",
	"newLine",
	"noEmit",
	"noEmitHelpers",
	"noEmitOnError",
	"outFile",
	"preserveConstEnums",
	"preserveValueImports",
	"removeComments",
	"sourceMap",
	"sourceRoot",
	"stripInternal",

	// Modules
	"allowArbitraryExtensions",
	"allowImportingTsExtensions",
	"allowUmdGlobalAccess",
	"baseUrl",
	"noResolve",
	"paths",
	"resolveJsonModule",
	"rootDirs",
	"typeRoots",
	"types",

	// Type checking
	"allowUnreachableCode",
	"allowUnusedLabels",
	"alwaysStrict",
	"exactOptionalPropertyTypes",
	"noFallthroughCasesInSwitch",
	"noImplicitAny",
	"noImplicitOverride",
	"noImplicitReturns",
	"noImplicitThis",
	"noPropertyAccessFromIndexSignature",
	"noUncheckedIndexedAccess",
	"noUnusedLocals",
	"noUnusedParameters",
	"strict",
	"strictBindCallApply",
	"strictFunctionTypes",
	"strictNullChecks",
	"strictPropertyInitialization",
	"useUnknownInCatchVariables",

	// JavaScript Support
	"allowJs",
	"checkJs",
	"maxNodeModuleJsDepth",

	// Projects
	"composite",
	"disableReferencedProjectLoad",
	"disableSolutionSearching",
	"disableSourceOfProjectReferenceRedirect",
	"incremental",
	"tsBuildInfoFile",

	// Editor Support
	"disableSizeLimit",
	"plugins",

	// InteropConstraints
	"allowSyntheticDefaultImports",
	"esModuleInterop",
	"forceConsistentCasingInFileNames",
	"isolatedModules",
	"preserveSymlinks",

	// Language and Environment
	"emitDecoratorMetadata",
	"experimentalDecorators",
	"jsx",
	"jsxFactory",
	"jsxFragmentFactory",
	"jsxImportSource",
	"lib",
	"noLib",
	"reactNamespace",
	"target",
	"useDefineForClassFields",

	// Diagnostics
	"diagnostics",
	"explainFiles",
	"extendedDiagnostics",
	"generateCpuProfile",
	"listEmittedFiles",
	"listFiles",
	"traceResolution",

	// Output formatting
	"noErrorTruncation",
	"preserveWatchOutput",
	"pretty",

	// Completeness
	"skipDefaultLibCheck",
	"skipLibCheck",

	// Watch Options
	"assumeChangesOnlyAffectDirectDependencies",

	// Backwards Compatibility
	"charset",
	"keyofStringsOnly",
	"noImplicitUseStrict",
	"noStrictGenericChecks",
	"out",
	"suppressExcessPropertyErrors",
	"suppressImplicitAnyIndexErrors",
] as const;

/**
 * Sorting order for tsconfig files.
 */
const sortOrder = [
	"extends",
	"include",
	"exclude",
	"compilerOptions",
	...compilerOptionsOrder,
	"references",
];

const orderMap: Map<string, number> = new Map();

for (const [index, key] of sortOrder.entries()) {
	orderMap.set(key, index);
}

/**
 * @beta
 */
export interface SortTsconfigResult {
	alreadySorted: boolean;
	tsconfig: string;
}

/**
 * Sorts a tsconfig file, optionally writing the changes back to the file.
 *
 * @param tsconfigPath - path to a tsconfig file
 * @param write - if true, the file will be overwritten with sorted content
 * @returns An object containing a boolean indicating whether the file was already sorted and the sorted tsconfig
 * string.
 *
 * @beta
 */
export function sortTsconfigFile(
	tsconfigPath: string,
	write: boolean,
): SortTsconfigResult {
	const sorted = isSorted(tsconfigPath);

	const origString = readFileSync(tsconfigPath).toString();
	const { indent } = detectIndent(origString);
	const sortedString = sortJsonc(origString, {
		sort: sortOrder,
		spaces: indent,
	});

	// normalize the original and sorted string using prettier so we can compare them safely

	if (!sorted && write) {
		writeFileSync(tsconfigPath, sortedString);
	}

	return {
		alreadySorted: sorted,
		tsconfig: sortedString,
	};
}

/**
 * Checks if a tsconfig file is sorted.
 *
 * @beta
 */
export function isSorted(tsconfig: string): boolean {
	// const { default: jsonc } = await JSONC;
	const content = readFileSync(tsconfig, { encoding: "utf8" });
	const currentValue: TsConfigJson = JSONC.parse(content);
	const result = objectIsSorted(currentValue);
	return result;
}

// biome-ignore lint/suspicious/noExplicitAny: <explanation>
function objectIsSorted(obj: { [s: string]: any }): boolean {
	const properties = [...Object.entries(obj)];
	let index = -1;
	for (const [key, value] of Object.entries(obj)) {
		index++;
		// const [key, value] = entry;

		// If the value is an object, recursively check the object's sort.
		if (isObject(value) && !objectIsSorted(value)) {
			return false;
		}

		const nextKey =
			index >= properties.length - 1 ? undefined : properties[index + 1][0];
		const sortIndex = getSortIndex(key);
		const nextSortIndex = getSortIndex(nextKey);
		if (sortIndex > nextSortIndex) {
			return false;
		}
	}
	return true;
}

// biome-ignore lint/suspicious/noExplicitAny: <explanation>
function isObject(value: any): boolean {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/strict-boolean-expressions, @typescript-eslint/no-unsafe-member-access
	return value && typeof value === "object" && value.constructor === Object;
}

function getSortIndex(key: string | undefined): number {
	// get the expected sort index of the key; if not found, (unexpected key) use a number greater than any sortIndex,
	// assuming those items will always be at the bottom
	return key === undefined
		? orderMap.size + 1
		: orderMap.get(key) ?? orderMap.size + 1;
}
