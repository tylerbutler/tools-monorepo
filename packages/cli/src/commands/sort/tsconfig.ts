import { existsSync, statSync } from "node:fs";
import path from "node:path";
import { Args, Flags } from "@oclif/core";
import { BaseCommand, isSorted, sortTsconfigFile } from "@tylerbu/cli-api";
import { globby } from "globby";

export default class SortTsconfig extends BaseCommand<typeof SortTsconfig> {
	static override readonly aliases = ["sort:tsconfigs"];

	static override readonly summary =
		"Sorts a tsconfig file in place, or check that one is sorted.";

	static override readonly description =
		"By default, the command will only check if a tsconfig is sorted. Use the --write flag to write the sorted contents back to the file.";

	static override readonly args = {
		tsconfig: Args.custom<string[]>({
			description:
				"Path to the tsconfig file to sort, or a glob path to select multiple tsconfigs.",
			required: true,
			parse: async (input) => {
				const patterns: string[] = [];
				if (existsSync(input)) {
					const stats = statSync(input);
					if (stats.isDirectory()) {
						patterns.push(path.join(input, "tsconfig.json"));
					}
				} else {
					patterns.push(input);
				}

				const results = await globby(patterns, { gitignore: true });
				return results ?? [];
			},
		})(),
	};

	static override readonly flags = {
		write: Flags.boolean({
			description:
				"Write the sorted contents back to the file. Without this flag, the command only checks that the file is sorted.",
			default: false,
		}),
		...BaseCommand.flags,
	};

	// biome-ignore lint/suspicious/useAwait: inherited method
	async run(): Promise<void> {
		const { tsconfig: tsconfigs } = this.args;
		const { write } = this.flags;

		if (tsconfigs === undefined) {
			this.error("No path or glob to tsconfigs provided.");
		}

		if (tsconfigs.length === 0) {
			this.error("No files found matching arguments");
		}

		const unsortedFiles: string[] = [];

		for (const tsconfig of tsconfigs) {
			if (write) {
				const result = sortTsconfigFile(tsconfig, write);
				this.log(
					result.alreadySorted
						? `File already sorted: ${tsconfig}`
						: `Wrote sorted file: ${tsconfig}`,
				);
			} else {
				const sorted = isSorted(tsconfig);
				if (!sorted) {
					unsortedFiles.push(tsconfig);
					this.errorLog(`Not sorted! ${tsconfig}`);
				}
			}
		}

		if (unsortedFiles.length > 0) {
			this.error(`Found ${unsortedFiles.length} unsorted files.`, {
				exit: 1,
			});
		}
	}
}
