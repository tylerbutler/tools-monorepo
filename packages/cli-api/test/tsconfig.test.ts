import { copyFile, readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { withDir } from "tmp-promise";
import { describe, expect, it } from "vitest";
import { isSorted, sortTsconfigFile } from "../src/tsconfig.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const testFiles = {
	sorted: "data/tsconfig.sorted.json",
	unsorted: "data/tsconfig.unsorted.json",
	unsortedUnknownKeys: "data/tsconfig.unsorted.unknown-keys.json",
};

describe("tsconfig", async () => {
	it("detects sorted", () => {
		const testFile = path.join(__dirname, testFiles.sorted);
		const result = isSorted(testFile);
		expect(result).to.be.true;
	});

	it("detects unsorted unknown keys", () => {
		const testFile = path.join(__dirname, testFiles.unsortedUnknownKeys);
		const result = isSorted(testFile);
		expect(result).to.be.false;
	});

	it("detects unsorted", () => {
		const testFile = path.join(__dirname, testFiles.unsorted);
		const result = isSorted(testFile);
		expect(result).to.be.false;
	});

	it("sortTsconfigFile", async () => {
		await withDir(
			async ({ path: tempDir }) => {
				const sourceFile = path.join(__dirname, testFiles.unsorted);
				const testFile = path.join(tempDir, path.basename(testFiles.unsorted));
				await copyFile(sourceFile, testFile);

				const result = sortTsconfigFile(testFile, true);
				expect(result.alreadySorted).to.be.false;
				expect(result).to.matchSnapshot();

				const rewritten = await readFile(testFile, { encoding: "utf8" });
				expect(result.tsconfig).to.equal(rewritten);
			},
			{
				// usafeCleanup ensures the cleanup doesn't fail if there are files in the directory
				unsafeCleanup: true,
			},
		);
	});

	it("sortTsconfigFile with unknown keys", async () => {
		await withDir(
			async ({ path: tempDir }) => {
				const sourceFile = path.join(__dirname, testFiles.unsortedUnknownKeys);
				const testFile = path.join(
					tempDir,
					path.basename(testFiles.unsortedUnknownKeys),
				);
				await copyFile(sourceFile, testFile);

				const result = sortTsconfigFile(testFile, true);
				expect(result.alreadySorted).to.be.false;
				expect(result).to.matchSnapshot();

				const rewritten = await readFile(testFile, { encoding: "utf8" });
				expect(result.tsconfig).to.equal(rewritten);
			},
			{
				// usafeCleanup ensures the cleanup doesn't fail if there are files in the directory
				unsafeCleanup: true,
			},
		);
	});
});
