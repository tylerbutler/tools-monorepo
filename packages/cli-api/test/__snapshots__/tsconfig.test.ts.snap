// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tsconfig > sortTsconfigFile 1`] = `
{
  "alreadySorted": false,
  "tsconfig": "{
	"include": [
		"src/**/*"
	],
	// Compiler options below.
	"compilerOptions": {
		"rootDir": "src",
		"outDir": "built",
		"module": "Node16",
		"moduleResolution": "Node16", // This is a comment to the right.
		"declaration": true,
		"importHelpers": true,
		"strict": true,
		// Composite projects are a way to split a project into smaller pieces that can be built together.
		"composite": true,
		"esModuleInterop": true,
		"target": "es2019"
	},
	"ts-node": {
		"esm": true
	}
}",
}
`;
